lms/djangoapps/django_comment_client/base/event_transformers.py

```diff
@@ -0,0 +1,111 @@
+"""
+Transformers for Discussion-related events.
+"""
+from django.contrib.auth.models import User
+from django.core.urlresolvers import reverse, NoReverseMatch
+
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.locator import CourseLocator
+
+from django_comment_client.base.views import add_truncated_title_to_event_data
+from django_comment_client.permissions import get_team
+from django_comment_client.utils import get_cached_discussion_id_map_by_course_id
+from track.transformers import EventTransformer, EventTransformerRegistry
+
+
+def _get_string(dictionary, key):
+    value = dictionary.get(key)
+    return (
+        value if isinstance(value, basestring)
+        else None
+    )
+
+
+@EventTransformerRegistry.register
+class ForumThreadViewedEventTransformer(EventTransformer):
+    """
+    Transformer to augment forum thread viewed Segment event from mobile apps
+    with fields that are either not available or not efficiently accessible
+    within the apps.
+    """
+
+    match_key = u'edx.forum.thread.viewed'
+
+    def process_event(self):
+        """
+        Transform incoming events.
+
+        For events from mobile, enhance with fields that are not available
+        within the apps.
+
+        Pass-through other events.
+        """
+        if 'event' not in self:
+            self['event'] = {}
```

Comment:
Yeah `EventTransformer`s are really weird. It's a subclass of `dict`, and adds members and methods to it.

What I'm doing here with `self['event'] = {}` is if an external application passes in an event without an `'event'` field, we add the `'event'` field and `event` member so that other transformers (which assume that `event` is a field and member) don't raise an exception.